x/programs/rust/examples/counter/src/lib.rs

```diff
@@ -47,20 +47,20 @@ fn inc_external(
     of: Address,
     amount: i64,
 ) -> i64 {
-    program.call_program(&target, max_units, "inc", &[Box::new(of), Box::new(amount)])
+    program.call_program(&target, max_units, "inc", serialize_params!(of, amount))
```

Comment:
imo I think serializing _every_ parameter passed to/from WASM would add nice simplicity to the code.

Reply:
Agreed. I think we talked about this offline, but just want to capture here for posterity. I actually think your declarative macro is nice. I would rename it `params` because it's both more terse, and I don't think the declarative macro should perform the serialization. 

Another thing we could do is pass in a tuple and generate a bunch of trait implementations for tuples of various sizes, that might even be better for DX (program-developer experience). Libraries like `Warp` do this [here](https://docs.rs/warp/latest/src/warp/generic.rs.html). It's a little crazy, but definitely doable. 

As for the actual serialization, it should be possible to implement a trait such that the primitives serialize to themselves and everything else serializes to a byte-slice? 

Anyway, not adding anything new here, I think you already have a good idea how to implement this.

Reply:
To recap what we talked about offline. 
- We are going to stay with the `params` declarative macro for now.
- All types are now being serialized. 
- In the future, to reduce the amount of serialization/deserialization, we will serialize all params once inside of a struct. Then we will pass the struct to `WASM` and deserialize it inside the `#[public]` proc macro